File: Python/textbook_track/chapter01/ba1n.py
# Generate the d-Neighborhood of a String
import sys

NUCLEOTIDES = ('A', 'C', 'G', 'T')


def calc_immediate_neighbours(pattern):
    neighbours = set()
    for ix, letter in enumerate(pattern):
        for nucleotide in NUCLEOTIDES:
            if nucleotide != letter:
                pattern_list = list(pattern)
                pattern_list[ix] = nucleotide
                neighbours.add(''.join(pattern_list))
    return neighbours


def generate_neighbourhood(text, d):
    neighbourhood = {text}
    for _ in range(d):
        neighbours_of_neighbours = set()
        for pattern in neighbourhood:
            immediate_neighbours = calc_immediate_neighbours(pattern)
            neighbours_of_neighbours.update(immediate_neighbours)
        neighbourhood.update(neighbours_of_neighbours)
    return neighbourhood


def main():
    data = sys.stdin.read().splitlines()
    pattern = data[0]
    d = int(data[1])
    result = generate_neighbourhood(pattern, d)
    for string in result:
        print(string)


if __name__ == '__main__':
    main()

File: Julia/textbook_track/chapter01/ba1n.jl
# Generate the d-Neighborhood of a String

const NUCLEOTIDES = ['A', 'C', 'G', 'T']


function calc_immediate_neighbours(pattern)
    neighbours = Set{String}()
    for (ix, letter) in enumerate(pattern)
        for nucleotide in NUCLEOTIDES
            if nucleotide != letter
                pattern_array = collect(pattern)
                pattern_array[ix] = nucleotide
                push!(neighbours, join(pattern_array))
            end
        end
    end
    neighbours
end


function generate_neighbourhood(text, d)
    neighbourhood = Set{String}([text])
    for _ in 1:d
        neighbours_of_neighbours = Set{String}()
        for pattern in neighbourhood
            immediate_neighbours = calc_immediate_neighbours(pattern)
            union!(neighbours_of_neighbours, immediate_neighbours)
        end
        union!(neighbourhood, neighbours_of_neighbours)
    end
    neighbourhood
end


function main()
    pattern = readline()
    d = parse(Int64, readline())
    result = generate_neighbourhood(pattern, d)
    foreach(println, result)
end


main()
